src/loaders/workspace_uploader/workspace_uploader.py

```diff
@@ -586,96 +642,156 @@ def _process_genome_objects(
     loader_helper.create_softlink_between_dirs(new_dir, target_dir)
 
 
-def _upload_assembly_files_in_parallel(
-    asu: AssemblyUtil,
-    ws: Workspace,
-    upload_env_key: str,
-    workspace_id: int,
-    load_id: str,
-    collections_source_dir: str,
-    wait_to_upload_objs: dict[str, str],
-    batch_size: int,
-    source_dir: str,
+def _process_batch_upload(
+        obj_tuples: list[WSObjTuple],
+        conf: Conf,
+        workspace_id: int,
+        load_id: str,
+        asu_client: AssemblyUtil,
+        gfu_client: GenomeFileUtil,
+        upload_assembly_only: bool = True,
+) -> list[UploadResult]:
+    if upload_assembly_only:
+        upload_results = _upload_assemblies_to_workspace(asu_client, workspace_id, load_id, obj_tuples)
+    else:
+        upload_results = _upload_genomes_to_workspace(gfu_client, workspace_id, load_id, obj_tuples, conf.job_data_dir)
+
+    return upload_results
+
+
+def _process_failed_uploads(
```

Comment:
IIUC it's possible here that the genome + assembly were uploaded, in which case they'd be added to the upload results returned from this method, but the assembly file didn't get linked, which would leave the data in an inconsistent state.

Reply:
I am not sure I follow you. the unlinked files will be process in `_post_process` right?

Reply:
The genome files will be, presumably, since they exist before the call begins. But if the call failed, that means the uploader never got the paths for the assemblies back from GFU, and never linked them into the source dir.

Reply:
in that case, you will need to rerun the script and it will be handled by `_post_process` right? It's exactly the same as Assembly only case. Maybe I am still not getting what you are trying ask.

Reply:
in future PRs, `_process_failed_uploads` and `_post_process` will be called first to handle any inconsistency state if that answers what you are asking.  `_process_failed_uploads` will check the existence of the FASTA file and download from linked genome, this will also happen in the future PRs.

Reply:
I'm not sure I understand what you're saying - the two comments above seem to contradict each other. _post_process only gets called after a successful upload or a failed upload, the latter case called with any genomes / assemblies found in the workspace that were part of the batch. It doesn't download any assembly files from the workspace AFAICT and doesn't have access to any assembly files saved by GFU because the file locations weren't returned. So without changes it'll run post processing, update the yaml  files, and the download will look successful when it isn't. 

IIUC the assembly files are only linked in _upload_genomes_to_workspace, which if the genbank_to_genomes call fails, will not allow the assembly files to be linked.

Reply:
If things are still confusing let's talk after the sprint planning meeting tomorrow

Reply:
in future PRs, for failed upload cases, `_process_failed_uploads` will download FASTA file from linked Assembly object. It gets the assembly ref from genome obj and calls `asu.get_assembly_as_fasta`.

Reply:
:+1: